features/202-session-management/api.md

```diff
@@ -0,0 +1,107 @@
+# Session Management APIs
+
+## SDK API
+```typescript
+// Field value would be empty string if not available
+interface UserAgent {
+    raw: string;
+    name: string; // browser name or app ID
+    version: string; // browser version or app version
+    os: string;
+    osVersion: string;
+    deviceName: string;
+    deviceModel: string;
+}
+
+interface Session {
+    id: string;
+    identityID: string;
+    createdAt: Date;
+    lastAccessedAt: Date;
+    createdByIP: string;
+    lastAccessedIP: string;
+    userAgent: UserAgent;
+    name: string;
+    data: object;
+}
+
+interface ExtraSessionInfoOptions {
+    collectDeviceName: boolean; // indicate SDK should collect device name information
+    deviceName?: string; // if falsy, SDK would try to detect it if possible
+}
+
+// list all sessions of current user
+function listSessions(): Promise<Session[]>;
+
+// get specific session information for current user
+// for master-key, can use for any user
+function getSession(sessionID: string): Promise<Session>;
+
+// update specific session name for current user
+// for master-key, can use for any user and also update custom attributes
+function updateSession(sessionID: string, patch: { name?: string; data?: JSONObject }): Promise<void>;
```

Comment:
@kiootic As discussed before, not sure if you remember -- it seems better to have "deviceName" customizable instead of "sessionName", it seems more useful for the use case, if someone need to be able to identify each devices, and prevent sessions being used on the wrong devices.

So please help to:

1. Remove `updateSession` it seems not useful (we can still keep it if we think there are use case for users to store `data`, but I don't have any in my mind yet)
2. Remove `name` in the `Session` Interface
3. Provide a client side SDK api to update `deviceName` manually.

Or maybe we should scrap the whole feature of a custom name for the initial release first.

Reply:
BTW, for custom attributes of sessions, would it be better to design it as:

1. Just remove `updateSession` API
2. Attributes in `X-Skygear-Extra-Info` header will be stored in the session custom data (we need to add a strict size limit for the extra session custom data to avoid spam, and also for now just keep the latest version's value of each key? Until we have the audit trail feature)

Reply:
- Currently, it supports both device name and session name. Device name can be set in code or discovered automatically by SDK, while Session name can be updated using API and persisted at server side.
- The main purpose of custom attributes is to store session specific trusted data (i.e. can be updated only with master key).
- The size limit of `X-Skygear-Extra-Info` header is already implemented. I consider it as an implementation details so I didn't write it into the spec.

Reply:
* Ic. How the device name is set in Client side SDK? Seems not covered in spec?
* What are the use cases of custom attributes? I'm thinking is it better to remove Session Name and Session data, if we don't have a use case. Avoid uncertain API interface for future troubles if we don't know what it is for.

Reply:
- It is set in `auth.extraSessionInfoOptions.deviceName`.
- I was thinking about traditional session data (e.g. `$_SESSION`). I don't think there are use case that strictly requires this custom attributes, so I could remove it for now.

Reply:
Noted, let's remove session data and `updateSession` until we got a real use cases (or know what we are handling lol)

Usability-wise, would it better if we change the option `collectDeviceName` into a function that try to automatically fill in `deviceName`? It would be less confusing in case someone manually set `deviceName`, when it will be overrided if `collectDeviceName` is true